### PR TITLE
Markdown

### PR DESCRIPTION
**Pull request by emord**
Paired with @orangejenny on this.

Should no longer have any of the changes of #407 in it.

Should put this through a UAT round.

cc: @czue

### CODE REVIEW

File: src/templates/markdown_help.html [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<a href="#" class="turn-markdown-off markdown-trigger">turn off markdown</a>
+<a href="#" class="turn-markdown-on markdown-trigger">turn on markdown</a>
+<a href="#" class="fd-help"
+            data-title=<%-title%>
+            data-content="CommCare supports a subset of the Markdown annotation language <br><br><a target='_blank' href='https://confluence.dimagi.com/display/TDH/Markdown'> See more</a>"
```
Comment (emord) [inline src/templates/markdown_help.html]: Link will need to change

File: src/javaRosa.js [hunk 2]
```diff
@@ -1055,6 +1054,85 @@ define([
...
+            parent.toggleClass("has-markdown");
+        };
+
+        markdownOutput = $('<div>').addClass("controls well markdown-output");
+
+        widget.handleChange = function() {
+            super_handleChange();
+            var val = widget.getValue(),
+                item = this.getItextItem();
+            if (/[~*#[\]]+|^-/m.test(val)) {
```
Comment (czue) [inline src/javaRosa.js]: maybe extract this test to a function (i think the intent is `looksLikeMarkdown(val)`?) so it's clear what the intent is.

Comment (czue) [inline src/javaRosa.js]: guessing the regex might get expanded in the future.

File: src/javaRosa.js [hunk 3]
```diff
@@ -1055,6 +1054,85 @@ define([
...
+                val = widget.getValue();
+
+            elem.detach('.markdown-output');
+            elem.append(markdownOutput);
+            elem.find('.control-label').append(markdown_help({title:options.lstring }));
+
+            markdownOff = elem.find('.turn-markdown-off').click(function() {
+                wantsMarkdown = false;
+                var item = widget.getItextItem();
+                item.hasMarkdown = false;
```
Comment (czue) [inline src/javaRosa.js]: stylistically it seems cleaner to just combine these two items and get rid of the variable.

File: tests/markdown.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
+/*jshint multistr: true */
```
Comment (czue) [inline tests/markdown.js]: :+1:

**Comment by czue** [comment]
took a cursory pass - generally looks good to me! will let @millerdev or @orangejenny do a full review and merge.

File: src/javaRosa.js [hunk 5]
```diff
@@ -1055,6 +1054,85 @@ define([
...
+                return false;
+            });
+
+            if (widget.getItextItem().hasMarkdown) {
+                parent.addClass("has-markdown");
+            }
+            else {
+                parent.addClass("markdown-ignorant");
+            }
+            if (/[-~*#[\]]+/.test(val)) {
```
Comment (millerdev) [inline src/javaRosa.js]: Is it intentional that this regex is different than the one above (currently line 1077)? I like @czue's idea of extracting it to a function.

File: src/javaRosa.js [hunk 6]
```diff
@@ -1613,6 +1694,13 @@ define([
                             xmlWriter.writeXML(xml.normalize(val));
                             xmlWriter.writeEndElement();
                         }
+                        if (writeMarkdown) {
```
Comment (millerdev) [inline src/javaRosa.js]: Would it work to change this to `if (item.hasMarkdown)` and eliminate the local variable?

File: src/javaRosa.js [hunk 7]
```diff
@@ -1613,6 +1694,13 @@ define([
...
+                            val = forms[0].getValueOrDefault(lang);
```
Comment (millerdev) [inline src/javaRosa.js]: `forms[0].getValueOrDefault(lang)` doesn't look right because you don't know which form `form[0]` is, but I could be mistaken. Are you looking for the default form? In that case you probably want `itext.get('default', lang)`, right?

File: src/util.js [hunk 8]
```diff
@@ -300,7 +304,11 @@ define([
                 "XML " + (opts.not ? "should not be equivalent" : "mismatch"));
         return patch;
     };
-        
+
+    that.markdownFull = function (text) {
```
Comment (millerdev) [inline src/util.js]: What do you think of renaming this to simply `markdown`?

File: tests/markdown.js (hunk 4 above)
Comment (millerdev) [inline tests/markdown.js]: wait, why is this needed? typically this was only necessary in old-style tests with inline XML strings.

File: tests/markdown.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    _,
+    form,
+    MARKDOWN_TEST_XML,
+    SIMPLE_MARKDOWN_XML,
+    SIMPLE_MARKDOWN_NO_CHARS_XML,
+    NO_MARKDOWN_XML,
+    NO_MARKDOWN_STARS_XML
+) {
+
+    // see note about controlling time in formdesigner.lock.js
```
Comment (millerdev) [inline tests/markdown.js]: is this comment relevant? I don't see any timing related code in this file.

File: tests/markdown.js [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+            it("should not write any markdown if markdown characters are not input", function() {
+                util.loadXML("");
+                util.addQuestion("Text", 'markdown_question');
+                $('[name=itext-en-label]').val("no markdown").change();
+                util.assertXmlEqual(call('createXML'), NO_MARKDOWN_XML, {normalize_xmlns: true});
+            });
+        });
+
+        describe("when a user explicitly wants no markdown", function() {
+            function beforeFn(done) {
```
Comment (millerdev) [inline tests/markdown.js]: Could use an anonymous function like `beforeEach(function (done) { ...`

File: tests/static/markdown/markdown-test.xml [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+			</instance>
+			<bind nodeset="/data/markdown_question" type="xsd:string" />
+			<bind nodeset="/data/no_markdown" type="xsd:string" />
+			<bind nodeset="/data/markdown_but_not" type="xsd:string" />
+			<bind nodeset="/data/escaped_markdown" type="xsd:string" />
+			<itext>
+				<translation lang="en" default="">
+					<text id="markdown_question-label">
+						<value>some markdown</value>
+                      	<value form="markdown">**some markdown**</value>
```
Comment (millerdev) [inline tests/static/markdown/markdown-test.xml]: inconsistent use of tabs and spaces for indentation

File: src/javaRosa.js (hunk 2 above)
Comment (emord) [inline src/javaRosa.js]: :+1: expanded and documented much better now

File: tests/markdown.js (hunk 4 above)
Comment (emord) [inline tests/markdown.js]: yeah. I just copied this file from javaRosa so I didn't need to do the boilerplate. cleaned this up

File: src/util.js (hunk 8 above)
Comment (emord) [inline src/util.js]: was originally based on copy-paste-dev so this was to differentiate with your markdownlite. I suppose that function could be thrown away on copy-paste once these are merged together.

File: src/javaRosa.js [hunk 12]
```diff
@@ -1055,6 +1054,95 @@ define([
...
+        var parent = options.parent;
+        var widget = itextLabelWidget(mug, language, form, options),
+            super_setValue = widget.setValue,
+            super_getUIElement = widget.getUIElement,
+            super_handleChange = widget.handleChange,
+            wantsMarkdown = true,
+            markdownOff, markdownOn, markdownOutput;
+
+        function looksLikeMarkdown(val) {
+            /* Regex checks (in order):
```
Comment (emord) [inline src/javaRosa.js]: Pulled these from: https://confluence.dimagi.com/display/TDH/Markdown

File: src/util.js (hunk 8 above)
Comment (millerdev) [inline src/util.js]: Yep, fine to remove `markdownlite` now that we have markdown-it.

**Comment by millerdev** [comment]
:+1:

**Comment by emord** [comment]
This passed QA, but I want to make sure the migration for REC is good to go before releasing

**Comment by millerdev** [comment]
Looks good. Ping when you're ready to merge.

**Comment by emord** [comment]
@millerdev I think it's good to merge now. Added a commit to always accept markdown as the true value and overwrite the default form. I confirmed with a/rec that always accepting markdown works for them
